JSON Schema to GBNF integration tests

HanClinto opened the pull request:
This adds a number of FAILING tests to exercise JSON schema conversion. These failures are good, as they show bugs in our current schema-to-grammar system (as noted by #7703 and #7789 ).

I originally started writing these tests to help me dig into the root causes behind #7703 , but I ran into some other issues along the way (hence #7789 ).

Of the failing tests, in particular, there are a number of "pass" test cases that are marked with TODO that SHOULD pass and currently do NOT.

The "fail" test cases that are improperly passing are not as big of a deal (these are things that we can't really expect to support, such as numeric minimums or ensuring uniqueness of array entries), so feel free to ignore those TODO entries.

This is meant as an aide to anyone else who wants to chip in on any of the json-schema features / bugs and wants an easier way to debug the json schema generation and test it against sample generations. Very helpfully the json-schema.org documentation often includes examples of matching and non-matching strings for example schemas in their documentation, and I have begun copying those into these integration tests directly.

HanClinto committed:
Adding simple bare-bones test for end-to-end integration test for json validation against auto-generated JSON-schema grammars.

HanClinto committed:
Adding additional examples as documented in #7789 . Also adding the ability to automatically output improperly failing grammars to debug output files so they can more easily be examined in the gbnf-validator program.

HanClinto committed:
Uncommenting formerly commented tests so that they fail for others who are attempting to reproduce the bugs.

HanClinto commented:
@ochafik Check out these latest tests -- I think you might like them. :)

ochafik commented:
Awesome! I went the other direction in https://github.com/ggerganov/llama.cpp/pull/7797 (integration testing the json w/ the rest), same difference :-)

I'll take a deeper look in the coming days!

HanClinto commented:
> Awesome! Was plotting something similar in #7797 :-)
> 
> I'll take a deeper look in the coming days!

Haha -- we went down nearly identical roads. Your framework is much nicer than mine -- I'll pull your framework into this branch and update my tests accordingly. Then you should be able to pull it into your branch more easily if you want. Nicely done!!

HanClinto committed:
Merging improved schema test methods added by @ochafik in #7797

HanClinto committed:
Adding #define to temporarily remove failing tests so that this PR can pass CI, but still be useful for other PRs that want to leverage the framework.

HanClinto commented:
Okay, ready for review / merge.

I incorporated the improvements from @ochafik 's test harness in #7797 , and also added a `#define INCLUDE_FAILING_TESTS 1` at the top that can be commented / uncommented to exclude / include the failing test cases. This is so that those test cases don't break CI, but the PR can still be merged so that other PRs can build on it in order to resolve / improve json-to-gbnf conversion (#7840, #7797, #7863, etc).

ochafik commented on tests/test-grammar-integration.cpp:
```diff
@@ -468,6 +496,532 @@ empty ::= "blah" | )""";
+        },
+        // Failing strings
+        {
+            "",
```
"\"\"",
"true"

ochafik commented on tests/test-grammar-integration.cpp:
```diff
@@ -468,6 +496,532 @@ empty ::= "blah" | )""";
+    );
+
+    test_schema(
+        "exotic formats (list)",
```
I would split "data", "uuid", "time" and "date-time" test cases

ochafik commented on tests/test-grammar-integration.cpp:
```diff
@@ -468,6 +496,532 @@ empty ::= "blah" | )""";
+        },
+        // Failing strings
+        {
+            "{}",
```
"\"",
"\"a",
"a\"",
"\"\\\"",

ochafik commented on tests/test-grammar-integration.cpp:
```diff
@@ -468,6 +496,532 @@ empty ::= "blah" | )""";
+            // "By extension, even an empty object is valid"
+            R"""({})""",
+            // "By default, providing additional properties is valid"
+#ifdef INCLUDE_FAILING_TESTS
```
Thanks for preparing these!

I would just move them down to the failed branch (w/ caveat header), hopefully I'll move them back up very soon :-) (not a fan of `#ifdef` dead code)

ochafik commented on tests/test-grammar-integration.cpp:
```diff
@@ -468,6 +496,532 @@ empty ::= "blah" | )""";
+        )""",
+        // Passing strings
+        {
+            "{\"b\": \"foo\", \"a\": \"bar\"}",
```
nit: TBH I'd be tempted to use R"""({...})""" as soon as any backslash is needed, to make it easier to copy-paste things.

ochafik commented on tests/test-grammar-integration.cpp:
```diff
@@ -468,6 +496,532 @@ empty ::= "blah" | )""";
+        {
+            "{}", // Missing all required properties
+            "{\"productName\": \"A green door\", \"price\": 12.50, \"productId\": 1}", // Out of order properties
+            // TODO: The following line should fail, but currently it passes. `exclusiveMinimum` is not supported, as it would likely be too difficult to implement.
```
nit: exclusiveMinimum for number is on my roadmap :-)

ochafik replied:
uncomment  & move it up / to passing?

ochafik reviewed (approved):
Amazing, thanks @HanClinto !

HanClinto committed:
Fixing nits from ochafik. Removing escape slashes, adding additional failing cases, fixing some other strings.

HanClinto committed:
Fixing grammar indentation to be consistent throughout file.

HanClinto commented:
@ochafik Thank you very much for the review!  I implemented most of your suggestions, but there were a couple that I didn't bother with (that I figure can get resolved in #7797 and others).
